google/cloud/security/scanner/scanners/groups_scanner.py

```diff
@@ -0,0 +1,168 @@
+                # Start at the tree root, find all instances of the specified
+                # group, then add the rule to all the members of the specified
+                # group.
+                for node in iterators.PreOrderIter(root):
+                    if node.member_email == rule.get('group_email'):
+                        node = self._append_rule(node, rule)
+
+        return self.find_violations(root)
+
+
+    def find_violations(self, root):  # pylint: disable=arguments-differ
+        """Find violations, starting from the given root.
+
+        At this point, we can start to find violations at each node!
+
+        We have a tree, with data populated at each node.
+        ...and rules are also applied at each node
+        Traversal order should not matter, since we need to evaluate all nodes.
+
+        Each node can have multiple rules.
+        Each rule can have multiple conditions
+
+        If a rule is violated, then the node is in violation.
+        i.e. if all rules pass, then the node is not in violation.
+
+        Args:
+            root: The list of policies to find violations in.
+
+        Returns:
+            A list of nodes that are in violation.
+        """
+        all_violations = []
+        for node in iterators.PreOrderIter(root):
+
+            # No need to evaluate these nodes.
+            # This represents the org, i.e. is not a group.
+            if node.member_email == 'my_customer':
+                continue
+            # This represents the auto-generated group, containing all the users
+            # in the org.
+            if node.member_email == '':
+                continue
+
+            node.violations = []
+            whitelist_rule_statuses = []
+            for rule in node.rules:
+                condition_statuses = []
+
+                if rule.get('mode') == 'whitelist':
+                    for condition in rule.get('conditions'):
+                        if condition.get('member_email') in node.member_email:
+                            condition_statuses.append(True)
+                        else:
+                            condition_statuses.append(False)
+
+                    # All the conditions of this rule have evaluated.
+                    # The rule is fulfilled, if any condition matches.
+                    if any(condition_statuses) is True:
+                        whitelist_rule_statuses.append(True)
+                    else:
+                        whitelist_rule_statuses.append(False)
+
+                elif rule.get('mode') == 'blacklist':
+                    pass  # TODO
+
+                elif rule.get('mode') == 'required':
+                    pass  # TODO
+
+                else:
+                    pass  # TODO
+    
+            # Determine if the node is in violations or not.
+            # All rules must be fulfilled, for a node to not be in violation.
+            # If any rule is not fulfilled, then node is in violation.
+            #
```

Comment:
I think this is really a nit, but the truth table is not necessary, given that someone who is writing python can just refer to the python official documentation for explanation of any()/all().